fs/smb/client/truenas_streams.c

```diff
@@ -0,0 +1,633 @@
+				OPEN_REPARSE_POINT,
+			.path = path,
+			.fid = fid_out,
+			.mode = ACL_NO_MODE,
+		};
+		switch (xattr_flags) {
+		case XATTR_CREATE:
+			oparms.disposition = FILE_CREATE;
+			break;
+		case XATTR_REPLACE:
+			oparms.disposition = FILE_OPEN;
+			break;
+		default:
+			// We may be able to optimize here by
+			// setting FILE_SUPERSEDE on open and avoid
+			// setting EOF as a separate operation.  
+			oparms.disposition = FILE_OPEN_IF;
+			break;
+		};
+		break;
+	case STREAM_OPEN_READ:
+		oparms = (struct cifs_open_parms) {
+			.tcon = tcon,
+			.cifs_sb = sb,
+			.desired_access = FILE_READ_ATTRIBUTES |
+				FILE_READ_DATA | SYNCHRONIZE,
+			.create_options = cifs_create_options(sb, CREATE_NOT_DIR) |
+				OPEN_REPARSE_POINT,
+			.disposition = FILE_OPEN,
+			.path = path,
+			.fid = fid_out,
+		};
+		break;
+	case STREAM_OPEN_DELETE:
+		oparms = (struct cifs_open_parms) {
+			.tcon = tcon,
+			.cifs_sb = sb,
+			.desired_access = DELETE | FILE_WRITE_ATTRIBUTES,
+			.create_options = cifs_create_options(sb, CREATE_NOT_DIR) |
+				OPEN_REPARSE_POINT | CREATE_DELETE_ON_CLOSE,
+			.disposition = FILE_OPEN,
+			.path = path,
+			.fid = fid_out,
+		};
+		break;
+	default:
+		BUG();
+	};
+
+	utf16_path = cifs_convert_stream_path_to_utf16(path, stream, sb);
+	if (!utf16_path) {
+		return -ENOMEM;
+	}
+
+	rc = SMB2_open(xid, &oparms, utf16_path, &oplock, info_out, NULL, NULL, NULL);
+	kfree(utf16_path);
+
+	return rc;
+}
+
+static int
+delete_stream(struct dentry *dentry, struct cifs_tcon *tcon,
+	      unsigned int xid, const char *path, const char *stream)
+{
+	int rc;
+	struct cifs_fid fid;
+	struct smb2_file_all_info info;
+
+	rc = do_stream_open(tcon, CIFS_SB(dentry->d_sb), STREAM_OPEN_DELETE, xid, path, stream, 0, &fid, &info);
+        if (rc) {
+		// To maintain consistency with removexattr
+		// failure modes, convert ENOENT to ENODATA.
+		if (rc == -ENOENT)
+			rc = -ENODATA;
+
+		return rc;
+        }
+
+	SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+	return rc;
+}
+
+static int
+write_stream(struct dentry *dentry, struct cifs_tcon *tcon,
+	     const char *data, size_t sz, int flags, unsigned int xid,
+	     const char *path, const char *stream)
+{
+	/*
+	 * Write alternate data stream for file specified by dentry.
+	 * The stream is temporarily opened to perform IO (there is
+	 * no open for stream associated in VFS).
+	 */
+	int rc;
+	struct cifs_fid fid;
+	struct cifs_io_parms io_parms;
+	struct kvec iov[2]; // header, data
+	struct smb2_file_all_info info;
+	unsigned int written = 0;
+	unsigned int total_written;
+	unsigned int wsize;
+
+	// We will potentially need to break up stream write into chunks.
+	// Typical IO size in SMB is 1 MiB and typical streams are much
+	// smaller than this.
+	wsize = tcon->ses->server->ops->wp_retry_size(d_inode(dentry));
+
+	// vfs_streams_xattr in Samba appends an extra NULL byte to stream
+	if (streams_samba_compat_enabled) {
+		sz -= streams_samba_compat_enabled;
+	}
```

Comment:
This bit of logic isn't that clear to me.  Could you elaborate on the comment please?

Are you saying that a samba server sitting **above** this client will write a trailing NULL that we want to strip off before sending over the wire?  Will anyone _other_ than samba ever write to this?

Besides, even if the code is correct IMO it'd improve readability to simply say:
```
	if (streams_samba_compat_enabled) {
		sz -= 1; /* Remove NULL that was appended by samba */
	}
```
(with whatever tweak to the comment necessary)

Reply:
That's a good point. I'll redirect user back to notes in the header file and make this clearer.